### PR TITLE
Vote extensions: base implementation

### LINKED ISSUE #9835: Sync vote extensions with Spec

### PR DESCRIPTION
**Pull request by sergio-mena**
Closes #9835

This PR is a mechanical backport of the first part of vote extensions from the `v0.36.x` branch.

This includes:
* #6646
* #6885
* #6815 / #7821
* #8031
* #8141

This PR is best reviewed by comparing each of these backported PRs to the corresponding commit in this PR (diff of diffs). The most important one is #8141 (the previous ones are the first implementation made in 2021 which predated the ABCI++ specification).

N.B: The changelog will be updated once branch `feature/abci++vef` is merged back into `main`

---

#### PR checklist

- [x] Tests written/updated, or no tests needed
- [x] `CHANGELOG_PENDING.md` updated, or no changelog entry needed
- [x] Updated relevant documentation (`docs/`) and code comments, or no
      documentation updates needed

### CODE REVIEW

File: test/e2e/app/app.go [hunk 1]
```diff
@@ -318,6 +392,70 @@ func (app *Application) ProcessProposal(_ context.Context, req *abci.RequestProc
...
+// without doing anything about them. In this case, it just makes sure that the
+// vote extension is a well-formed integer value.
+func (app *Application) VerifyVoteExtension(_ context.Context, req *abci.RequestVerifyVoteExtension) (*abci.ResponseVerifyVoteExtension, error) {
+	// We allow vote extensions to be optional
+	if len(req.VoteExtension) == 0 {
+		return &abci.ResponseVerifyVoteExtension{
+			Status: abci.ResponseVerifyVoteExtension_ACCEPT,
+		}, nil
+	}
+	if req.Height != int64(app.state.Height)+1 {
```
Comment (williambanfield) [inline test/e2e/app/app.go:435]: Is this updated later? It seems wrong to me to have the app REJECT in this case.

**Review by williambanfield — APPROVED** [review]
Reviewed with the understanding that https://github.com/tendermint/tendermint/issues/9396 is also going to be cherry-picked, this looks all pretty simple and correct.

**Comment by sergio-mena** [comment]
> Reviewed with the understanding that #9396 is also going to be cherry-picked, this looks all pretty simple and correct.

Yes, all bullet points in #9396 are to be cherry-picked

File: consensus/msgs_test.go [hunk 2]
```diff
@@ -347,6 +347,7 @@ func TestConsMsgsVectors(t *testing.T) {
 		Timestamp:        date,
 		Type:             tmproto.PrecommitType,
 		BlockID:          bi,
+		Extension:        []byte("extension"),
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply consensus/msgs_test.go:350]: Just checked and this is not done in any of the upcoming cherry-picked commits. Introducing it now

File: test/e2e/app/app.go (hunk 1 above)
Comment (sergio-mena) [reply test/e2e/app/app.go:435]: Thanks for catching this! Just checked and this stays this way after upcoming patches.
Turning this into a panic here

**Pull request merged**